src/main/hook/patch/gfx.c

```diff
@@ -88,6 +101,63 @@ void wait_frame_limit(unsigned int fps){
     last_swap_time = current_time;
 }
 
+void calculate_zoom_factors(int src_width, int src_height, int dest_width, int dest_height, float *zoom_x, float *zoom_y) {
+    *zoom_x = (float) dest_width / (float) src_width;
+    *zoom_y = (float) dest_height / (float) src_height;
+}
+
+// Check if our window dimensions changed and update our settings.
+void GetCurrentWindowDimensions(Display *dpy,GLXDrawable drawable){
+    Window root;
+    int x, y;
+    unsigned int width, height, border_width, depth;
+    XGetGeometry(dpy, drawable, &root, &x, &y, &width, &height, &border_width, &depth);
+	if(width != s_target_display_width || height != s_target_display_height){		
+		s_target_display_width = width;
+		s_target_display_height = height;
+		// Recalculate Zoom Factor
+		calculate_zoom_factors(s_init_display_width,s_init_display_height,s_target_display_width,s_target_display_height,&s_zoom_factor_x,&s_zoom_factor_y);
+	}
+}
+
+// A modified version of the S3DResize function from the engine to support multimode adjustment.
+void S3DResizeEx(void){
+  if(s_scaling_mode == PATCH_GFX_SCALE_MODE_STRETCH){
+    glViewport(0, 0, s_target_display_width, s_target_display_height);
+  }else if(s_scaling_mode == PATCH_GFX_SCALE_MODE_PIXEL_PERFECT){
+    float targetAspectRatio = s_init_display_width / (float)s_init_display_height;
+
+    // Figure out the largest area that fits in this resolution at the desired aspect ratio
+    int width = s_target_display_width;
+    int height = (int)(width / targetAspectRatio + 0.5f);
+
+    if (height > s_target_display_height) {
+        // It doesn't fit our height, we must switch to pillarbox then
+        height = s_target_display_height;
+        width = (int)(height * targetAspectRatio + 0.5f);
+    }
+
+    // Calculate pillarbox dimensions
+    int pillarbox_x = (s_target_display_width - width) / 2;
+    int pillarbox_y = (s_target_display_height - height) / 2;
+    if(width != s_target_display_width || height != s_target_display_height){
+      s_target_display_width = width;
+      s_target_display_height = height;
+      calculate_zoom_factors(s_init_display_width,s_init_display_height,s_target_display_width,s_target_display_height,&s_zoom_factor_x,&s_zoom_factor_y);
+    }
+
+    
+    // Set up the new viewport with pillarboxing
+    glViewport(pillarbox_x, pillarbox_y, width, height);
+
+  }
+    glMatrixMode(GL_PROJECTION);
+    glLoadIdentity();
+    glOrtho(0.0f, s_init_display_width, 0.0f, s_init_display_height, -500.0f, 500.0f);
+    glMatrixMode(GL_MODELVIEW);
+    glLoadIdentity();  
```

Comment:
We know it's pump, but this is a very dynamic way of implementing the scaling compared to scaling the final output to a larger framebuffer. Mind adding a comment or two to explain why you decided for that and what are potentially issues with that compared to framebuffer scaling (if that is even an applicable solution here)?